### PR TITLE
Feat: Update new version for zingmp3

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

In the future 'mp3.zing.vn' will be disabled and upgraded to 'zingmp3.vn' with new feature and UI. And all the method to get data, and endpoint of API will be changed. This PR will update the function to get data from 'mp3.zing.vn' and 'zingmp3.vn'.

### CODE REVIEW

File: youtube_dl/extractor/zingmp3.py [hunk 1]
```diff
@@ -1,23 +1,52 @@
...
 from ..utils import (
-    ExtractorError,
     int_or_none,
+    try_get,
 )
 
 
 class ZingMp3BaseIE(InfoExtractor):
-    _VALID_URL_TMPL = r'https?://(?:mp3\.zing|zingmp3)\.vn/(?:%s)/[^/]+/(?P<id>\w+)\.html'
+    _VALID_URL_TMPL = r'https?://(?:mp3\.zing|zingmp3)\.vn/(?P<type>(?:%s))/[^/]+/(?P<id>\w+)(?:\.html|\?)'
```
Comment (dirkf) [inline youtube_dl/extractor/zingmp3.py:18]: As the `<id>` match will end on `[.#?]`, do you need the `.html` match at all?

File: youtube_dl/extractor/zingmp3.py [hunk 2]
```diff
@@ -1,23 +1,52 @@
...
+        'bai-hat': '/api/v2/page/get/song',
+        'embed': '/api/v2/page/get/song',
+        'video-clip': '/api/v2/page/get/video',
+        'playlist': '/api/v2/page/get/playlist',
+        'album': '/api/v2/page/get/playlist',
+        'lyric': '/api/v2/lyric/get/lyric',
+        'song_streaming': '/api/v2/song/get/streaming',
+    }
+
+    _TIMESTAMP = str(int(datetime.datetime.now().timestamp()))
```
Comment (dirkf) [inline youtube_dl/extractor/zingmp3.py:32]: `compat_str(...)` if you want it to be unicode text.

File: youtube_dl/extractor/zingmp3.py [hunk 3]
```diff
@@ -36,61 +65,100 @@ def _extract_item(self, item, fatal):
...
+        mobj = re.match(self._VALID_URL, url)
+
+        song_id = mobj.group('id')
+        type_url = mobj.group('type')
+
+        api = self.get_api_with_signature(name_api=self._SLUG_API[type_url], param={
+            'ctime': self._TIMESTAMP,
+            'id': song_id})
+        info = self.download_json(api, song_id)
+        return self._process_data(info.get('data'), song_id, type_url)
```
Comment (dirkf) [inline youtube_dl/extractor/zingmp3.py:122]: `_process_data(x,...)` expects `x` to be a `dict`. The downloaded 'JSON' may turn out to be a list or just `False`. So:
```
        info = self.download_json(api, song_id)
        info_data = try_get(info, lambda x: x['data'], dict)
        if not info_data:
            # if the site might often do this, not just when the API changes, add
            # `, expected=True`
            raise ExtractorError('API returned empty or invalid song data')
        return self._process_data(info_data, song_id, type_url)
```

File: youtube_dl/extractor/zingmp3.py [hunk 4]
```diff
@@ -1,23 +1,52 @@
...
+        'song_streaming': '/api/v2/song/get/streaming',
+    }
+
+    _TIMESTAMP = str(int(datetime.datetime.now().timestamp()))
+    _API_KEY = '88265e23d4284f25963e6eedac8fbfa3'
+    _SECRET_KEY = b'2aa2d1c561e809b267f3638c4a307aab'
+
+    def _extract_item(self, item, song_id, type_url, fatal):
+        item_id = item.get('encodeId') or song_id
+        title = item.get('title') or item.get('alias')
```
Comment (dirkf) [inline youtube_dl/extractor/zingmp3.py:38]: `title` is required, so use the original pattern `item.get('title') or item['alias']`

Or if you want to make the ensuing crash report clearer, `dict_get(item, ('title', 'alias')) or item['title']`

File: youtube_dl/extractor/zingmp3.py (hunk 2 above)
Comment (hatienl0i261299) [reply youtube_dl/extractor/zingmp3.py:32]: I'm thinking this timestamp is just a string to put in the hash, so maybe I'll hardcode it a sample text or remove it.

File: youtube_dl/extractor/zingmp3.py (hunk 4 above)
Comment (hatienl0i261299) [reply youtube_dl/extractor/zingmp3.py]: Yes, i updated

File: youtube_dl/extractor/zingmp3.py (hunk 3 above)
Comment (hatienl0i261299) [reply youtube_dl/extractor/zingmp3.py]: Updated.

File: youtube_dl/extractor/zingmp3.py (hunk 1 above)
Comment (hatienl0i261299) [reply youtube_dl/extractor/zingmp3.py]: Updated, ths you.

**Review by dirkf — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Comment by dirkf** [comment]
Would you still like this PR to be progressed?

**Comment by hatienl0i261299** [comment]
> Would you still like this PR to be progressed?

I opened this PR for a long time ago, and i fixed all the code review from you, but this PR is still not merged. And i think you dont need to merge this PR, and i closed it.
But if you still want to use this PR, i will sync with latest code from yt-dlp and re-open PR again. Thanks you.

**Pull request closed without merging**